File: PROCESSING/NatureOfCode/2_Forces/ForceMover/src/Sketch.java
import processing.core.PApplet;
import processing.core.PVector;

import java.util.ArrayList;

public class Sketch extends PApplet {

	public static void main(String[] args) {
		PApplet.main("Sketch");
	}
	
	
	class Mover {
		
		PVector p;// position
		PVector v; // velocity
		PVector a; // acceleration
		
		Mover(float x, float y) {
			p = new PVector(x, y);
			v = new PVector();
			a = new PVector();
		}
		
		void move() {
			
			
			v.add(a);
			p.add(v);
			
			a.mult(0); // clear acceleration
		}
		
		void draw() {
			pushStyle();
			fill(50);
			ellipse(p.x, p.y, 30, 30);
			popStyle();
		}
		
		void applyForce(PVector force) {
			a.add(force);
		}
		
	}
	
	class Baloon extends Mover {
		
		Baloon(float x, float y) {
			super(x,y);
		}
		
		@Override void move() {
			PVector helium = new PVector(0, -0.05f);

			applyForce(helium);

			if (p.y < 15) {
				v.y *= -1;
			}
			
			if (p.x < 15 || p.x > width-15) {
				v.x *= -1;
			}
			
			super.move();
		}
		
	}
	
	ArrayList<Baloon> baloons = new ArrayList<>();
	
	float xoff = 0;

	@Override public void settings() {
		size(800, 400);
	}

	@Override public void setup() {
		surface.setTitle("Baluuunz xddd");
		for (int i = 0; i<20; i++) {
			baloons.add(new Baloon(random(width), 100 + random(height-100)));
		}
	}

	@Override public void draw() {
		background(200);

		PVector wind = new PVector(0.05f*(2*noise(xoff)-1), 0);
		xoff += 0.01f;
		
		for (Baloon b : baloons) {
			b.applyForce(wind);
			b.move();
			
			line(b.p.x, b.p.y, b.p.x, b.p.y + 50);
			b.draw();
		}
		
	}
}

File: PROCESSING/NatureOfCode/9_GeneticAlgorithms/SmartRockets/src/Sketch.kt
import processing.core.PApplet
import processing.core.PVector

class Sketch : PApplet() {

	inner class RocketPopulation(val initPos : PVector,
	                             val populationSize : Int,
	                             val mutationRate : Float,
	                             val rocketLifetime : Int,
	                             val target : PVector) {
		
		var members = Array(populationSize) {
			// ACHTUNG -> pos random
			Rocket(PVector(random(150f, 250f), random(150f, 250f)), rocketLifetime)
		}
		
		var matingPool = ArrayList<Rocket>()
		var generations = 0
		
		fun generate() {
			
			// perform selection according to target fitness
			for (member in members) {
				member.fit(target)
			}
			
			// select to mating pool
			matingPool.clear()
			for (member in members) {
				var n = (member.fitness * populationSize).toInt()

				for (j in 0 until n) {
					matingPool.add(member) // add references of member according to fitness
				}
			}

			// reproduce
			for ( i in members.indices ) {
				// reproduction
				// select indices
				var a = random(matingPool.size.toFloat()).toInt()
				var b = random(matingPool.size.toFloat()).toInt()

				// get parents
				// perhaps check if parents not same object ? ( gains not that much efficiency )
				var parentA = matingPool[a]
				var parentB = matingPool[b]

				var child = Rocket(initPos, rocketLifetime)
				child.dna = parentA.crossover(parentB.dna)
				
				// mutate child
				child.mutate(mutationRate)

				members[i] = child
			}
			
			generations++
			
		}
		
		fun live() {
			for (m in members) {
				m.run()
			}
		}
		
	}
	
	inner class DNA(lifetime : Int) {
		var maxforce = 0.1f
		var genes = Array<PVector>(lifetime) {
			PVector.random2D().mult(random(0f, maxforce))
		}
	}
	
	inner class Rocket(var pos : PVector,
	                   val lifetime : Int) {
		var vel = PVector()
		var acc = PVector()
		
		var dna = DNA(lifetime)
		var fitness = 0f
		
		var geneCounter = 0
		
		fun applyForce(f : PVector) {
			acc.add(f)
		}
		
		fun update() {
			vel.add(acc)
			pos.add(vel)
			acc.mult(0f)
		}
		
		fun draw() {
			pushStyle()
			fill(0)
			ellipse(pos.x, pos.y, 2f, 2f)
			popStyle()
		}
		
		fun run() {
			applyForce(dna.genes[geneCounter])
			geneCounter++
			update()
		}
		
		fun fit(target : PVector) {
			var d = PVector.dist(pos, target)
			fitness = 1/d
		}

		fun crossover(partner : DNA) : DNA {
			var child = DNA(lifetime)

			var midpoint = random(dna.genes.size.toFloat()).toInt()

			for ( i in dna.genes.indices ) {
				if (i > midpoint) 	child.genes[i] = dna.genes[i]
				else 				child.genes[i] = partner.genes[i]
			}

			return child
		}

		fun mutate(rate : Float) {
			for ( i in dna.genes.indices ) {
				if (random(1f) < rate) {
					dna.genes[i] = PVector.random2D().mult(random(0f, dna.maxforce))
				}
			}
		}
		
	}
	
	
	val LIFETIME = 200
	var lifeCounter = 0
	
	lateinit var population : RocketPopulation

	override fun settings() {
		size(400, 400)
	}

	override fun setup() {
		population = RocketPopulation(PVector(200f, 200f), 10, 0.01f, LIFETIME, PVector(100f, 100f))
	}

	override fun draw() {
		
		background(200)
		
		if (lifeCounter < LIFETIME ) {
			population.live()
			lifeCounter++
		} else {
			lifeCounter = 0
			population.generate()
		}
		
		for (rocket in population.members) {
			rocket.draw()
		}
		
	}
	
}

fun main(args: Array<String>) {
	PApplet.main("Sketch")
}